cmd/collectors/zapi/plugins/shelf/shelf.go

```diff
@@ -280,3 +280,149 @@ func (my *Shelf) Run(data *matrix.Matrix) ([]*matrix.Matrix, error) {
+	//shelves = result.SearchChildren([]string{"shelf-environ-channel-info", "shelf-environ-shelf-list", "shelf-environ-shelf-info"})
+
+	if len(channels) == 0 {
+		return nil, errors.New(errors.ERR_NO_INSTANCE, "no channels found")
+	}
+
+	var output []*matrix.Matrix
+
+	// Purge and reset data
+	for _, data1 := range my.data {
+		data1.PurgeInstances()
+		data1.Reset()
+	}
+
+	for _, channel := range channels {
+		channelName := channel.GetChildContentS("channel-name")
+		shelves = channel.SearchChildren([]string{"shelf-environ-shelf-list", "shelf-environ-shelf-info"})
+
+		if len(shelves) == 0 {
+			my.Logger.Warn().Str("channel", channelName).Msg("no shelves found")
+			continue
+		}
+
+		for _, shelf := range shelves {
+
+			uid := shelf.GetChildContentS("shelf-id")
+			shelfName := uid // no shelf name in 7mode
+			shelfId := uid
+
+			for attribute, data1 := range my.data {
+				if statusMetric := data1.GetMetric("status"); statusMetric != nil {
+
+					if my.instanceKeys[attribute] == "" {
+						my.Logger.Warn().Str("attribute", attribute).Msg("no instance keys defined")
+						continue
+					}
+
+					objectElem := shelf.GetChildS(attribute)
+					if objectElem == nil {
+						my.Logger.Warn().Str("attribute", attribute).Msg("no instances on this system")
+						continue
+					}
+
+					my.Logger.Debug().Msgf("fetching %d [%s] instances", len(objectElem.GetChildren()), attribute)
+
+					for _, obj := range objectElem.GetChildren() {
+
+						if key := obj.GetChildContentS(my.instanceKeys[attribute]); key != "" {
+							instanceKey := shelfId + "." + key + "." + channelName
+							instance, err := data1.NewInstance(instanceKey)
+
+							if err != nil {
+								my.Logger.Error().Msgf("add (%s) instance: %v", attribute, err)
+								return nil, err
+							}
+							my.Logger.Debug().Msgf("add (%s) instance: %s.%s", attribute, shelfId, key)
+
+							for label, labelDisplay := range my.instanceLabels[attribute].Map() {
+								if value := obj.GetChildContentS(label); value != "" {
+									instance.SetLabel(labelDisplay, value)
+								}
+							}
+
+							instance.SetLabel("shelf", shelfName)
+							instance.SetLabel("shelf_id", shelfId)
+							instance.SetLabel("channel", channelName)
+
+							// Each child would have different possible values which is ugly way to write all of them,
+							// so normal value would be mapped to 1 and rest all are mapped to 0.
+							if instance.GetLabel("status") == "normal" {
+								statusMetric.SetValueInt(instance, 1)
+							} else {
+								statusMetric.SetValueInt(instance, 0)
+							}
+
+						} else {
+							my.Logger.Debug().Msgf("instance without [%s], skipping", my.instanceKeys[attribute])
+						}
+					}
+
+					output = append(output, data1)
+				}
+			}
+		}
+	}
+
+	// second loop to populate numeric data
+
+	for _, channel := range channels {
+		channelName := channel.GetChildContentS("channel-name")
```

Comment:
do we want to complete channel loop and again do iterate? or can we do 2 for loop of shelves in same channel loop?

Reply:
done